RNN Multi GPU error

**Describe the bug**
When trying to use multiple GPUs to train RNN models, I get the following error: 

> AssertionError: Caught AssertionError in replica 1 on device 1.
Original Traceback (most recent call last):
  File "/home/students/kiegeland/anaconda3/envs/nmt/lib/python3.7/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
    output = module(*input, **kwargs)
  File "/home/students/kiegeland/anaconda3/envs/nmt/lib/python3.7/site-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/students/kiegeland/test/joeynmt/joeynmt/model.py", line 87, in forward
    trg_mask=kwargs["trg_mask"])
  File "/home/students/kiegeland/test/joeynmt/joeynmt/model.py", line 147, in _encode_decode
    trg_mask=trg_mask)
  File "/home/students/kiegeland/test/joeynmt/joeynmt/model.py", line 186, in _decode
    trg_mask=trg_mask)
  File "/home/students/kiegeland/anaconda3/envs/nmt/lib/python3.7/site-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/students/kiegeland/test/joeynmt/joeynmt/decoders.py", line 340, in forward
    prev_att_vector=prev_att_vector)
  File "/home/students/kiegeland/test/joeynmt/joeynmt/decoders.py", line 192, in _check_shapes_input_forward
    assert src_mask.shape[2] == encoder_output.shape[1]
AssertionError

It seems like the src_length is different for each device which results in different shapes for the encoder output.
Is there anything else I need to configure?

**To Reproduce**
Training the reverse experiment using 2 gpus on version=1.0
[node_configuration.txt](https://github.com/joeynmt/joeynmt/files/5493901/train_reverse_sh.txt)

**Logged output**
[error_log.txt](https://github.com/joeynmt/joeynmt/files/5493824/reverse_error.txt)

--- a/joeynmt/batch.py
+++ b/joeynmt/batch.py
@@ -3,13 +3,14 @@
 """
 Implementation of a mini-batch.
 """
+import torch
 
 
 class Batch:
     """Object for holding a batch of data with mask during training.
     Input is a batch from a torch text iterator.
     """
-
+    # pylint: disable=too-many-instance-attributes
     def __init__(self, torch_batch, pad_index, use_cuda=False):
         """
         Create a new joey batch from a torch batch.
@@ -30,6 +31,7 @@ class Batch:
         self.trg_length = None
         self.ntokens = None
         self.use_cuda = use_cuda
+        self.device = torch.device("cuda" if self.use_cuda else "cpu")
 
         if hasattr(torch_batch, "trg"):
             trg, trg_length = torch_batch.trg
@@ -42,7 +44,7 @@ class Batch:
             self.trg_mask = (self.trg_input != pad_index).unsqueeze(1)
             self.ntokens = (self.trg != pad_index).data.sum().item()
 
-        if use_cuda:
+        if self.use_cuda:
             self._make_cuda()
 
     def _make_cuda(self):
@@ -51,14 +53,14 @@ class Batch:
 
         :return:
         """
-        self.src = self.src.cuda()
-        self.src_mask = self.src_mask.cuda()
-        self.src_length = self.src_length.cuda()
+        self.src = self.src.to(self.device)
+        self.src_mask = self.src_mask.to(self.device)
+        self.src_length = self.src_length.to(self.device)
 
         if self.trg_input is not None:
-            self.trg_input = self.trg_input.cuda()
-            self.trg = self.trg.cuda()
-            self.trg_mask = self.trg_mask.cuda()
+            self.trg_input = self.trg_input.to(self.device)
+            self.trg = self.trg.to(self.device)
+            self.trg_mask = self.trg_mask.to(self.device)
 
     def sort_by_src_length(self):
         """

--- a/joeynmt/decoders.py
+++ b/joeynmt/decoders.py
@@ -185,8 +185,9 @@ class RecurrentDecoder(Decoder):
         :param prev_att_vector:
         """
         assert len(encoder_output.shape) == 3
-        assert len(encoder_hidden.shape) == 2
-        assert encoder_hidden.shape[-1] == encoder_output.shape[-1]
+        if encoder_hidden is not None:
+            assert len(encoder_hidden.shape) == 2
+            assert encoder_hidden.shape[-1] == encoder_output.shape[-1]
         assert src_mask.shape[1] == 1
         assert src_mask.shape[0] == encoder_output.shape[0]
         assert src_mask.shape[2] == encoder_output.shape[1]
@@ -306,18 +307,18 @@ class RecurrentDecoder(Decoder):
          initialize the first hidden decoder state
          (when `self.init_hidden_option` is "bridge" or "last").
 
-        :param trg_embed: emdedded target inputs,
+        :param trg_embed: embedded target inputs,
             shape (batch_size, trg_length, embed_size)
         :param encoder_output: hidden states from the encoder,
             shape (batch_size, src_length, encoder.output_size)
         :param encoder_hidden: last state from the encoder,
-            shape (batch_size x encoder.output_size)
+            shape (batch_size, encoder.output_size)
         :param src_mask: mask for src states: 0s for padded areas,
             1s for the rest, shape (batch_size, 1, src_length)
-        :param unroll_steps: number of steps to unrol the decoder RNN
+        :param unroll_steps: number of steps to unroll the decoder RNN
         :param hidden: previous decoder hidden state,
             if not given it's initialized as in `self.init_hidden`,
-            shape (num_layers, batch_size, hidden_size)
+            shape (batch_size, num_layers, hidden_size)
         :param prev_att_vector: previous attentional vector,
             if not given it's initialized with zeros,
             shape (batch_size, 1, hidden_size)
@@ -329,6 +330,19 @@ class RecurrentDecoder(Decoder):
             - att_vectors: attentional vectors
                 with shape (batch_size, unroll_steps, hidden_size)
         """
+        # initialize decoder hidden state from final encoder hidden state
+        if hidden is None and encoder_hidden is not None:
+            hidden = self._init_hidden(encoder_hidden)
+        else:
+            # DataParallel splits batch along the 0th dim.
+            # Place back the batch_size to the 1st dim here.
+            if isinstance(hidden, tuple):
+                h, c = hidden
+                hidden = (h.permute(1, 0, 2).contiguous(),
+                          c.permute(1, 0, 2).contiguous())
+            else:
+                hidden = hidden.permute(1, 0, 2).contiguous()
+            # shape (num_layers, batch_size, hidden_size)
 
         # shape checks
         self._check_shapes_input_forward(
@@ -338,10 +352,6 @@ class RecurrentDecoder(Decoder):
             src_mask=src_mask,
             hidden=hidden,
             prev_att_vector=prev_att_vector)
-
-        # initialize decoder hidden state from final encoder hidden state
-        if hidden is None:
-            hidden = self._init_hidden(encoder_hidden)
 
         # pre-compute projected encoder outputs
         # (the "keys" for the attention mechanism)
@@ -378,6 +388,18 @@ class RecurrentDecoder(Decoder):
         # att_probs: batch, unroll_steps, src_length
         outputs = self.output_layer(att_vectors)
         # outputs: batch, unroll_steps, vocab_size
+
+        # DataParallel gathers batches along the 0th dim.
+        # Put batch_size dim to the 0th position.
+        if isinstance(hidden, tuple):
+            h, c = hidden
+            hidden = (h.permute(1, 0, 2).contiguous(),
+                      c.permute(1, 0, 2).contiguous())
+            assert hidden[0].size(0) == batch_size
+        else:
+            hidden = hidden.permute(1, 0, 2).contiguous()
+            assert hidden.size(0) == batch_size
+        # shape (batch_size, num_layers, hidden_size)
         return outputs, hidden, att_probs, att_vectors
 
     def _init_hidden(self, encoder_final: Tensor = None) \

--- a/joeynmt/encoders.py
+++ b/joeynmt/encoders.py
@@ -51,7 +51,6 @@ class RecurrentEncoder(Encoder):
         :param freeze: freeze the parameters of the encoder during training
         :param kwargs:
         """
-
         super().__init__()
 
         self.emb_dropout = torch.nn.Dropout(p=emb_dropout, inplace=False)
@@ -87,8 +86,8 @@ class RecurrentEncoder(Encoder):
         assert len(src_length.shape) == 1
 
     #pylint: disable=arguments-differ
-    def forward(self, embed_src: Tensor, src_length: Tensor, mask: Tensor) \
-            -> (Tensor, Tensor):
+    def forward(self, embed_src: Tensor, src_length: Tensor, mask: Tensor,
+                **kwargs) -> (Tensor, Tensor):
         """
         Applies a bidirectional RNN to sequence of embeddings x.
         The input mini-batch x needs to be sorted by src length.
@@ -109,6 +108,7 @@ class RecurrentEncoder(Encoder):
         self._check_shapes_input_forward(embed_src=embed_src,
                                          src_length=src_length,
                                          mask=mask)
+        total_length = embed_src.size(1)
 
         # apply dropout to the rnn input
         embed_src = self.emb_dropout(embed_src)
@@ -121,7 +121,8 @@ class RecurrentEncoder(Encoder):
         if isinstance(hidden, tuple):
             hidden, memory_cell = hidden
 
-        output, _ = pad_packed_sequence(output, batch_first=True)
+        output, _ = pad_packed_sequence(output, batch_first=True,
+                                        total_length=total_length)
         # hidden: dir*layers x batch x hidden
         # output: batch x max_length x directions*hidden
         batch_size = hidden.size()[1]
@@ -195,7 +196,7 @@ class TransformerEncoder(Encoder):
     def forward(self,
                 embed_src: Tensor,
                 src_length: Tensor,
-                mask: Tensor) -> (Tensor, Tensor):
+                mask: Tensor, **kwargs) -> (Tensor, Tensor):
         """
         Pass the input (and mask) through each layer in turn.
         Applies a Transformer encoder to sequence of embeddings x.

--- a/joeynmt/model.py
+++ b/joeynmt/model.py
@@ -124,7 +124,9 @@ class Model(nn.Module):
                                                       src_length=src_length,
                                                       src_mask=src_mask,
                                                       **kwargs)
+
         unroll_steps = trg_input.size(1)
+        assert "decoder_hidden" not in kwargs.keys()
         return self._decode(encoder_output=encoder_output,
                             encoder_hidden=encoder_hidden,
                             src_mask=src_mask, trg_input=trg_input,
@@ -141,7 +143,8 @@ class Model(nn.Module):
         :param src_mask:
         :return: encoder outputs (output, hidden_concat)
         """
-        return self.encoder(self.src_embed(src), src_length, src_mask)
+        return self.encoder(self.src_embed(src), src_length, src_mask,
+                            **_kwargs)
 
     def _decode(self, encoder_output: Tensor, encoder_hidden: Tensor,
                 src_mask: Tensor, trg_input: Tensor,
@@ -168,7 +171,8 @@ class Model(nn.Module):
                             unroll_steps=unroll_steps,
                             hidden=decoder_hidden,
                             prev_att_vector=att_vector,
-                            trg_mask=trg_mask)
+                            trg_mask=trg_mask,
+                            **_kwargs)
 
     def __repr__(self) -> str:
         """

--- a/joeynmt/prediction.py
+++ b/joeynmt/prediction.py
@@ -240,7 +240,7 @@ def parse_test_args(cfg, mode="test"):
     tokenizer_info = f"[{sacrebleu['tokenize']}]" \
         if eval_metric == "bleu" else ""
 
-    return batch_size, batch_type, use_cuda, n_gpu, level, \
+    return batch_size, batch_type, use_cuda, device, n_gpu, level, \
            eval_metric, max_output_length, beam_size, beam_alpha, \
            postprocess, bpe_type, sacrebleu, decoding_description, \
            tokenizer_info
@@ -288,7 +288,7 @@ def test(cfg_file,
         trg_vocab = datasets["trg_vocab"]
 
     # parse test args
-    batch_size, batch_type, use_cuda, n_gpu, level, eval_metric, \
+    batch_size, batch_type, use_cuda, device, n_gpu, level, eval_metric, \
         max_output_length, beam_size, beam_alpha, postprocess, \
         bpe_type, sacrebleu, decoding_description, tokenizer_info \
         = parse_test_args(cfg, mode="test")
@@ -301,7 +301,7 @@ def test(cfg_file,
     model.load_state_dict(model_checkpoint["model_state"])
 
     if use_cuda:
-        model.cuda()
+        model.to(device)
 
     # multi-gpu eval
     if n_gpu > 1 and not isinstance(model, torch.nn.DataParallel):
@@ -433,7 +433,7 @@ def translate(cfg_file: str, ckpt: str, output_path: str = None) -> None:
     src_field.vocab = src_vocab
 
     # parse test args
-    batch_size, batch_type, use_cuda, n_gpu, level, _, \
+    batch_size, batch_type, use_cuda, device, n_gpu, level, _, \
         max_output_length, beam_size, beam_alpha, postprocess, \
         bpe_type, sacrebleu, _, _ = parse_test_args(cfg, mode="translate")
 
@@ -445,7 +445,7 @@ def translate(cfg_file: str, ckpt: str, output_path: str = None) -> None:
     model.load_state_dict(model_checkpoint["model_state"])
 
     if use_cuda:
-        model.cuda()
+        model.to(device)
 
     if not sys.stdin.isatty():
         # input file given

--- a/joeynmt/prediction.py
+++ b/joeynmt/prediction.py
@@ -201,8 +201,11 @@ def parse_test_args(cfg, mode="test"):
     device = torch.device("cuda" if use_cuda else "cpu")
     if mode == 'test':
         n_gpu = torch.cuda.device_count() if use_cuda else 0
-        logger.info("Process device: %s, n_gpu: %d, batch_size per device: %d",
-            device, n_gpu, batch_size // n_gpu if n_gpu > 1 else batch_size)
+        k = cfg["testing"].get("beam_size", 1)
+        batch_per_device = batch_size*k // n_gpu if n_gpu > 1 else batch_size*k
+        logger.info("Process device: %s, n_gpu: %d, "
+                    "batch_size per device: %d (with beam_size)",
+                    device, n_gpu, batch_per_device)
         eval_metric = cfg["training"]["eval_metric"]
 
     elif mode == 'translate':

--- a/joeynmt/search.py
+++ b/joeynmt/search.py
@@ -197,17 +197,23 @@ def beam_search(model: Model, size: int,
     transformer = isinstance(model.decoder, TransformerDecoder)
     batch_size = src_mask.size(0)
     att_vectors = None  # not used for Transformer
+    hidden = None       # not used for Transformer
+    trg_mask = None     # not used for RNN
 
     # Recurrent models only: initialize RNN hidden state
     # pylint: disable=protected-access
     if not transformer:
+        # tile encoder states and decoder initial states beam_size times
         hidden = model.decoder._init_hidden(encoder_hidden)
-    else:
-        hidden = None
-
-    # tile encoder states and decoder initial states beam_size times
-    if hidden is not None:
         hidden = tile(hidden, size, dim=1)  # layers x batch*k x dec_hidden_size
+        # DataParallel splits batch along the 0th dim.
+        # Place back the batch_size to the 1st dim here.
+        if isinstance(hidden, tuple):
+            h, c = hidden
+            hidden = (h.permute(1, 0, 2), c.permute(1, 0, 2))
+        else:
+            hidden = hidden.permute(1, 0, 2)
+            # batch*k x layers x dec_hidden_size
 
     encoder_output = tile(encoder_output.contiguous(), size,
                           dim=0)  # batch*k x src_len x enc_hidden_size
@@ -219,8 +225,6 @@ def beam_search(model: Model, size: int,
         if isinstance(model, torch.nn.DataParallel):
             trg_mask = torch.stack(
                 [src_mask.new_ones([1, 1]) for _ in model.device_ids])
-    else:
-        trg_mask = None
 
     # numbering elements in the batch
     batch_offset = torch.arange(batch_size, dtype=torch.long, device=device)
@@ -273,7 +277,7 @@ def beam_search(model: Model, size: int,
             logits, hidden, att_scores, att_vectors = model(
                 return_type="decode",
                 encoder_output=encoder_output,
-                encoder_hidden=encoder_hidden,
+                encoder_hidden=None, # used to initialize decoder_hidden only
                 src_mask=src_mask,
                 trg_input=decoder_input, #trg_embed = embed(decoder_input)
                 decoder_hidden=hidden,
@@ -384,12 +388,12 @@ def beam_search(model: Model, size: int,
             if isinstance(hidden, tuple):
                 # for LSTMs, states are tuples of tensors
                 h, c = hidden
-                h = h.index_select(1, select_indices)
-                c = c.index_select(1, select_indices)
+                h = h.index_select(0, select_indices)
+                c = c.index_select(0, select_indices)
                 hidden = (h, c)
             else:
                 # for GRUs, states are single tensors
-                hidden = hidden.index_select(1, select_indices)
+                hidden = hidden.index_select(0, select_indices)
 
         if att_vectors is not None:
             att_vectors = att_vectors.index_select(0, select_indices)

--- a/joeynmt/training.py
+++ b/joeynmt/training.py
@@ -462,9 +462,9 @@ class TrainManager:
         # get loss
         batch_loss, _, _, _ = self.model(return_type="loss", **vars(batch))
 
-        # average on multi-gpu parallel training
+        # sum multi-gpu losses
         if self.n_gpu > 1:
-            batch_loss = batch_loss.mean()
+            batch_loss = batch_loss.sum()
 
         # normalize batch loss
         if self.normalization == "batch":
@@ -479,6 +479,9 @@ class TrainManager:
                 "or summation of loss 'none' implemented")
 
         norm_batch_loss = batch_loss / normalizer
+
+        if self.n_gpu > 1:
+            norm_batch_loss = norm_batch_loss / self.n_gpu
 
         if self.batch_multiplier > 1:
             norm_batch_loss = norm_batch_loss / self.batch_multiplier

--- a/joeynmt/training.py
+++ b/joeynmt/training.py
@@ -169,7 +169,7 @@ class TrainManager:
         self.n_gpu = torch.cuda.device_count() if self.use_cuda else 0
         self.device = torch.device("cuda" if self.use_cuda else "cpu")
         if self.use_cuda:
-            self.model.cuda()
+            self.model.to(self.device)
 
         # fp16
         self.fp16 = train_config.get("fp16", False)
@@ -300,7 +300,7 @@ class TrainManager:
 
         # move parameters to cuda
         if self.use_cuda:
-            self.model.cuda()
+            self.model.to(self.device)
 
         # fp16
         if self.fp16 and model_checkpoint.get("amp_state", None) is not None:

--- a/test/unit/test_decoder.py
+++ b/test/unit/test_decoder.py
@@ -193,20 +193,19 @@ class TestRecurrentDecoder(TensorTestCase):
         self.assertEqual(output.shape, torch.Size(
             [batch_size, time_dim, self.vocab_size]))
         self.assertEqual(hidden.shape, torch.Size(
-            [self.num_layers, batch_size, self.hidden_size]))
+            [batch_size, self.num_layers, self.hidden_size]))
         self.assertEqual(att_probs.shape, torch.Size(
             [batch_size, time_dim, time_dim]))
         self.assertEqual(att_vectors.shape, torch.Size(
             [batch_size, time_dim, self.hidden_size]))
         hidden_target = torch.Tensor(
-            [[[0.1814, 0.5468, -0.4717, -0.7580, 0.5834, -0.4018],
-             [0.1814, 0.5468, -0.4717, -0.7580, 0.5834, -0.4018]],
+            [[[ 0.1814,  0.5468, -0.4717, -0.7580,  0.5834, -0.4018],
+              [ 0.4649,  0.5484, -0.2702,  0.4545,  0.1983,  0.2771],
+              [-0.1752, -0.4215,  0.1941, -0.3975, -0.2317, -0.5566]],
 
-            [[0.4649, 0.5484, -0.2702, 0.4545, 0.1983, 0.2771],
-             [0.4649, 0.5484, -0.2702, 0.4545, 0.1983, 0.2771]],
-
-            [[-0.1752, -0.4215, 0.1941, -0.3975, -0.2317, -0.5566],
-             [-0.1752, -0.4215, 0.1941, -0.3975, -0.2317, -0.5566]]])
+             [[ 0.1814,  0.5468, -0.4717, -0.7580,  0.5834, -0.4018],
+              [ 0.4649,  0.5484, -0.2702,  0.4545,  0.1983,  0.2771],
+              [-0.1752, -0.4215,  0.1941, -0.3975, -0.2317, -0.5566]]])
         output_target = torch.Tensor(
             [[[ 0.2702, -0.1988, -0.1985, -0.2998, -0.2564],
              [ 0.2719, -0.2075, -0.2017, -0.2988, -0.2595],

